File: CS/App_Code/CriteraVisitorBase.cs
using DevExpress.Data.Filtering;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

public class CriteraVisitorBase : IClientCriteriaVisitor<CriteriaOperator> {
    public CriteraVisitorBase() { }

    #region IClientCriteriaVisitor Members

    public virtual CriteriaOperator Visit(JoinOperand theOperand) {
        CriteriaOperator condition = theOperand.Condition.Accept(this) as CriteriaOperator;
        CriteriaOperator expression = theOperand.AggregatedExpression.Accept(this) as CriteriaOperator;
        if (object.ReferenceEquals(condition, null) || object.ReferenceEquals(expression, null)) return null;
        return new JoinOperand(theOperand.JoinTypeName, condition, theOperand.AggregateType, expression);
    }

    public virtual CriteriaOperator Visit(OperandProperty theOperand) {
        return theOperand;
    }

    public virtual CriteriaOperator Visit(AggregateOperand theOperand) {
        OperandProperty operand = theOperand.CollectionProperty.Accept(this) as OperandProperty;
        CriteriaOperator condition = theOperand.Condition.Accept(this) as CriteriaOperator;
        CriteriaOperator expression = theOperand.AggregatedExpression.Accept(this) as CriteriaOperator;
        if (object.ReferenceEquals(condition, null) || object.ReferenceEquals(expression, null) || object.ReferenceEquals(operand, null)) return null;
        return new AggregateOperand(operand, expression, theOperand.AggregateType, condition);
    }
    #endregion

    #region ICriteriaVisitor Members

    public virtual CriteriaOperator Visit(FunctionOperator theOperator) {
        List<CriteriaOperator> operators = new List<CriteriaOperator>();
        foreach (CriteriaOperator op in theOperator.Operands) {
            CriteriaOperator temp = op.Accept<CriteriaOperator>(this);
            if (object.ReferenceEquals(temp, null)) return null;
            operators.Add(temp);
        }
        return new FunctionOperator(theOperator.OperatorType, operators);
    }

    public virtual CriteriaOperator Visit(OperandValue theOperand) {
        return theOperand;
    }

    public virtual CriteriaOperator Visit(GroupOperator theOperator) {
        List<CriteriaOperator> operators = new List<CriteriaOperator>();
        foreach (CriteriaOperator op in theOperator.Operands) {
            CriteriaOperator temp = op.Accept<CriteriaOperator>(this);
            if (object.ReferenceEquals(temp, null)) continue;
            operators.Add(temp);
        }
        return new GroupOperator(theOperator.OperatorType, operators);
    }

    public virtual CriteriaOperator Visit(InOperator theOperator) {
        CriteriaOperator leftOperand = theOperator.LeftOperand.Accept<CriteriaOperator>(this);
        List<CriteriaOperator> operators = new List<CriteriaOperator>();
        foreach (CriteriaOperator op in theOperator.Operands) {
            CriteriaOperator temp = op.Accept<CriteriaOperator>(this);
            if (object.ReferenceEquals(temp, null)) continue;
            operators.Add(temp);
        }
        if (object.ReferenceEquals(leftOperand, null)) return null;
        return new InOperator(leftOperand, operators);
    }

    public virtual CriteriaOperator Visit(UnaryOperator theOperator) {
        CriteriaOperator operand = theOperator.Operand.Accept<CriteriaOperator>(this);
        if (object.ReferenceEquals(operand, null)) return null;
        return new UnaryOperator(theOperator.OperatorType, operand);
    }

    public virtual CriteriaOperator Visit(BinaryOperator theOperator) {
        CriteriaOperator leftOperand = theOperator.LeftOperand.Accept<CriteriaOperator>(this);
        CriteriaOperator rightOperand = theOperator.RightOperand.Accept<CriteriaOperator>(this);
        if (object.ReferenceEquals(leftOperand, null) || object.ReferenceEquals(rightOperand, null)) return null;
        return new BinaryOperator(leftOperand, rightOperand, theOperator.OperatorType);
    }

    public virtual CriteriaOperator Visit(BetweenOperator theOperator) {
        CriteriaOperator test = theOperator.TestExpression.Accept<CriteriaOperator>(this);
        CriteriaOperator begin = theOperator.BeginExpression.Accept<CriteriaOperator>(this);
        CriteriaOperator end = theOperator.EndExpression.Accept<CriteriaOperator>(this);
        if (object.ReferenceEquals(test, null) || object.ReferenceEquals(begin, null) || object.ReferenceEquals(end, null)) return null;
        return new BetweenOperator(test, begin, end);
    }
    #endregion
}

File: VB/App_Code/CriteraVisitorBase.vb
Imports DevExpress.Data.Filtering
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web

Public Class CriteraVisitorBase
	Implements IClientCriteriaVisitor(Of CriteriaOperator)

	Public Sub New()
	End Sub

	#Region "IClientCriteriaVisitor Members"

	Public Overridable Function Visit(ByVal theOperand As JoinOperand) As CriteriaOperator Implements IClientCriteriaVisitor(Of CriteriaOperator).Visit
		Dim condition As CriteriaOperator = TryCast(theOperand.Condition.Accept(Me), CriteriaOperator)
		Dim expression As CriteriaOperator = TryCast(theOperand.AggregatedExpression.Accept(Me), CriteriaOperator)
		If Object.ReferenceEquals(condition, Nothing) OrElse Object.ReferenceEquals(expression, Nothing) Then
			Return Nothing
		End If
		Return New JoinOperand(theOperand.JoinTypeName, condition, theOperand.AggregateType, expression)
	End Function

	Public Overridable Function Visit(ByVal theOperand As OperandProperty) As CriteriaOperator Implements IClientCriteriaVisitor(Of CriteriaOperator).Visit
		Return theOperand
	End Function

	Public Overridable Function Visit(ByVal theOperand As AggregateOperand) As CriteriaOperator Implements IClientCriteriaVisitor(Of CriteriaOperator).Visit
		Dim operand As OperandProperty = TryCast(theOperand.CollectionProperty.Accept(Me), OperandProperty)
		Dim condition As CriteriaOperator = TryCast(theOperand.Condition.Accept(Me), CriteriaOperator)
		Dim expression As CriteriaOperator = TryCast(theOperand.AggregatedExpression.Accept(Me), CriteriaOperator)
		If Object.ReferenceEquals(condition, Nothing) OrElse Object.ReferenceEquals(expression, Nothing) OrElse Object.ReferenceEquals(operand, Nothing) Then
			Return Nothing
		End If
		Return New AggregateOperand(operand, expression, theOperand.AggregateType, condition)
	End Function
	#End Region

	#Region "ICriteriaVisitor Members"

	Public Overridable Function Visit(ByVal theOperator As FunctionOperator) As CriteriaOperator Implements DevExpress.Data.Filtering.ICriteriaVisitor(Of CriteriaOperator).Visit
		Dim operators As New List(Of CriteriaOperator)()
		For Each op As CriteriaOperator In theOperator.Operands
			Dim temp As CriteriaOperator = op.Accept(Of CriteriaOperator)(Me)
			If Object.ReferenceEquals(temp, Nothing) Then
				Return Nothing
			End If
			operators.Add(temp)
		Next op
		Return New FunctionOperator(theOperator.OperatorType, operators)
	End Function

	Public Overridable Function Visit(ByVal theOperand As OperandValue) As CriteriaOperator Implements DevExpress.Data.Filtering.ICriteriaVisitor(Of CriteriaOperator).Visit
		Return theOperand
	End Function

	Public Overridable Function Visit(ByVal theOperator As GroupOperator) As CriteriaOperator Implements DevExpress.Data.Filtering.ICriteriaVisitor(Of CriteriaOperator).Visit
		Dim operators As New List(Of CriteriaOperator)()
		For Each op As CriteriaOperator In theOperator.Operands
			Dim temp As CriteriaOperator = op.Accept(Of CriteriaOperator)(Me)
			If Object.ReferenceEquals(temp, Nothing) Then
				Continue For
			End If
			operators.Add(temp)
		Next op
		Return New GroupOperator(theOperator.OperatorType, operators)
	End Function

	Public Overridable Function Visit(ByVal theOperator As InOperator) As CriteriaOperator Implements DevExpress.Data.Filtering.ICriteriaVisitor(Of CriteriaOperator).Visit
		Dim leftOperand As CriteriaOperator = theOperator.LeftOperand.Accept(Of CriteriaOperator)(Me)
		Dim operators As New List(Of CriteriaOperator)()
		For Each op As CriteriaOperator In theOperator.Operands
			Dim temp As CriteriaOperator = op.Accept(Of CriteriaOperator)(Me)
			If Object.ReferenceEquals(temp, Nothing) Then
				Continue For
			End If
			operators.Add(temp)
		Next op
		If Object.ReferenceEquals(leftOperand, Nothing) Then
			Return Nothing
		End If
		Return New InOperator(leftOperand, operators)
	End Function

	Public Overridable Function Visit(ByVal theOperator As UnaryOperator) As CriteriaOperator Implements DevExpress.Data.Filtering.ICriteriaVisitor(Of CriteriaOperator).Visit
		Dim operand As CriteriaOperator = theOperator.Operand.Accept(Of CriteriaOperator)(Me)
		If Object.ReferenceEquals(operand, Nothing) Then
			Return Nothing
		End If
		Return New UnaryOperator(theOperator.OperatorType, operand)
	End Function

	Public Overridable Function Visit(ByVal theOperator As BinaryOperator) As CriteriaOperator Implements DevExpress.Data.Filtering.ICriteriaVisitor(Of CriteriaOperator).Visit
		Dim leftOperand As CriteriaOperator = theOperator.LeftOperand.Accept(Of CriteriaOperator)(Me)
		Dim rightOperand As CriteriaOperator = theOperator.RightOperand.Accept(Of CriteriaOperator)(Me)
		If Object.ReferenceEquals(leftOperand, Nothing) OrElse Object.ReferenceEquals(rightOperand, Nothing) Then
			Return Nothing
		End If
		Return New BinaryOperator(leftOperand, rightOperand, theOperator.OperatorType)
	End Function

	Public Overridable Function Visit(ByVal theOperator As BetweenOperator) As CriteriaOperator Implements DevExpress.Data.Filtering.ICriteriaVisitor(Of CriteriaOperator).Visit
		Dim test As CriteriaOperator = theOperator.TestExpression.Accept(Of CriteriaOperator)(Me)
		Dim begin As CriteriaOperator = theOperator.BeginExpression.Accept(Of CriteriaOperator)(Me)
		Dim [end] As CriteriaOperator = theOperator.EndExpression.Accept(Of CriteriaOperator)(Me)
		If Object.ReferenceEquals(test, Nothing) OrElse Object.ReferenceEquals(begin, Nothing) OrElse Object.ReferenceEquals([end], Nothing) Then
			Return Nothing
		End If
		Return New BetweenOperator(test, begin, [end])
	End Function
	#End Region
End Class